android/src/main/java/com/reactnativeteadssdkmodule/RNInReadAdPlacement.kt

```diff
@@ -0,0 +1,198 @@
+package com.reactnativeteadssdkmodule
+
+import android.os.Handler
+import android.os.Looper
+import android.util.Log
+import com.facebook.react.bridge.*
+import com.facebook.react.modules.core.DeviceEventManagerModule
+import tv.teads.sdk.*
+import tv.teads.sdk.renderer.InReadAdView
+import java.util.*
+
+
+class RNInReadAdPlacement(reactContext: ReactApplicationContext) :
+  ReactContextBaseJavaModule(reactContext) {
+
+  override fun getName(): String {
+    return "RNInReadAdPlacement"
+  }
+
+  private fun sendEvent(reactContext: ReactContext, eventName: String, params: WritableMap?) {
+    reactContext
+      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
+      .emit(eventName, params)
+  }
+
+  //requestAd
+  @ReactMethod
+  fun requestAd(pid: Int, settingsMap: ReadableMap, promise: Promise) {
+    val instanceIdentifier = UUID.randomUUID().toString()
+    RNAdInstanceManager.instance(pid).inReadAdPlacement.requestAd(AdRequestSettings.fromMap(
+      settingsMap.toHashMap().toMap()
+    ),
+      object : InReadAdListener {
+        override fun adOpportunityTrackerView(trackerView: AdOpportunityTrackerView) {
+          val params = Arguments.createMap()
+          params.putString("didPlay", "didPlay")
+          sendEvent(reactApplicationContext, "didPlay", params)
+        }
+
+        override fun onAdReceived(
+          inReadAdView: InReadAdView,
+          adRatio: AdRatio
+        ) {
+          RNAdInstanceManager.new(
+            RNAdInstanceManager.AdInstanceMap(
+              inReadAdView,
+              instanceIdentifier
+            )
+          )
+          promise.resolve(instanceIdentifier)
+          val params = Arguments.createMap()
+          params.putString("didReceiveAd", "didReceiveAd")
+          sendEvent(reactApplicationContext, "didReceiveAd", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d("from ad", "didReceiveAd")
+
+          }
+        }
+
+        override fun onAdClicked() {
+          val params = Arguments.createMap()
+          params.putString("didRecordClick", "didRecordClick")
+          sendEvent(reactApplicationContext, "didRecordClick", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d("form ad", "didRecordClick")
+          }
+        }
+
+        override fun onAdClosed() {
+          val params = Arguments.createMap()
+          params.putString("didClose", "didClose")
+          sendEvent(reactApplicationContext, "didClose", params)
+          Handler(Looper.getMainLooper()).post {
+            Log.d(
+              "from ad",
+              "didClose"
+            )
+          }
+        }
+
+        override fun onAdError(code: Int, description: String) {
+          val params = Arguments.createMap()
```

Comment:
So when you do this `val params = Arguments.createMap()` it returns the current params from the function, in that case `(code: Int, description: String)`?

Reply:
You are right there were no details in the response from the event. 
I've added the code and description in the event `didCatchError`, now the publisher implementing the listener for `didCatchError` will be able to see, the adId corresponding to the ad, and the code and description from the error.